---
format:
  html:
    toc: false      # hide TOC in HTML
  pdf:
    toc: true       # keep TOC in PDF
---

::: {.content-visible when-format="html"}

<div class="paper-authors">
  <p><strong>Nina Dombrowski</strong></p>
  <p>Institute for Biodiversity and Ecosystem Dynamics (IBED), University of Amsterdam (UVA)</p>
  <p><a href="mailto:n.dombrowski@uva.nl">n.dombrowski@uva.nl</a></p>
</div>

:::

# Introduction {.unnumbered} 

In this tutorial, you will learn how to work on the command line and use a High-Performance Computing (HPC) environment. This will allow you to analyse 16S long-reads and turn this data into a count table of microbial taxa.

This course is designed for students in microbial ecology with little or no prior experience using the command line. During this tutorial, you will learn how to:

1. **Install a Terminal**

2. **Document your Code**
    - Decide what software to use for code documentation
    - How to use markdown and document workflows

3. **Navigate the Command Line**
    - Navigate the filesystem
    - Work with files
    - Search and filter data
    - View and understand sequence files

4. **Understand the setup of an HPC**

5. **Work with an HPC**
    - Connect to an HPC
    - Understand job schedulers (e.g., SLURM)
    - Run command-line tools (e.g., `fastqc`, `seqkit`)
    - Set up and activating conda environments

6. **Analyse long-read data**
    - Analyse sequence read quality
    - Perform quality filtering of sequence read data
    - Align quality-filtered reads to a reference database 
    - Generate a count table


::: {.content-hidden}

```{bash}
# Standalone rendering 
quarto render --profile standalone source/terminal.qmd
quarto render --profile standalone source/documentation.qmd
quarto render --profile standalone source/hpc_intro.qmd
```

::: {.callout-question .callout-warning collapse=false}
# Question 

X

::: {.callout-answer .callout-warning collapse=true}
# Click to see the answer 

X

:::
:::


:::

</div>