---
format:
  html:
    toc: false      # hide TOC in HTML
  pdf:
    toc: true       # keep TOC in PDF
---

::: {.content-visible when-format="html"}

<div class="paper-authors">
  <p><strong>Nina Dombrowski</strong></p>
  <p>Institute for Biodiversity and Ecosystem Dynamics (IBED), University of Amsterdam (UVA)</p>
  <p><a href="mailto:n.dombrowski@uva.nl">n.dombrowski@uva.nl</a></p>
</div>

:::

# Introduction {.unnumbered} 

Welcome to the *Microbial Ecology Data Analysis Course*!  

In this short tutorial, you will learn how to use the command line and a High-Performance Computing (HPC) environment in order to move from raw 16S nanopore reads to a basic count table of microbial taxa.

This course is designed for students in microbial ecology with little or no prior experience using the command line. During this tutorial, you will learn how to:

1. **Install a Terminal**
    - How to access the command line 
    - How to run basic commands

2. **Document your Code**
    - Decide what software to use for code documentation
    - How to use markdown and document workflows

3. **Navigate the Command Line**
    - Navigate the filesystem (`cd`, `ls`, `pwd`)
    - Work with files (`mkdir`, `cp`, `mv`, `rm`, `cat`, `head`, `less`)
    - Search and filter data (`grep`, `wc`, `cut`)
    - View and underrstand FASTQ files

4. **Work with an HPC**
    - Connect to an HPC (SSH)
    - Understand job schedulers (e.g., SLURM)
    - Run command-line tools (e.g., `fastqc`, `seqkit`)
    - Set up and activating conda environments
    - Run workflows with Snakemake


::: {.content-hidden}

::: {.callout-question .callout-warning collapse=false}
# Question 

X

::: {.callout-answer .callout-warning collapse=true}
# Click to see the answer 

X

:::
:::


:::

</div>