---
title: |
    Introduction to Data Analysis

    in Microbial Ecology

author: "Nina Dombrowski"
logo: "../img/logo.png"
format: 
    revealjs: 
        theme: [slides.scss, slides.css]
        transition: fade
        slide-number: true
        style: "text-align: right; left: 10%"
        preview-links: true
        code-annotations: below
        highlight-style: "vim-dark"

engine: knitr
embed-resources: true

revealjs-plugins:
    - pointer
    - editable
    - qrcode
filters:
  - editable
---

## Introduction to Bioinformatics

[Bioinformatics applies computational methods to store, manage, and analyse biological data]{.highlighted}

- Experiments involving sequencing generates too much data to analyse manually
- Your Nanopore sequencing generated ~ 100,000 sequencing reads
- Computational tools allow you to store, filter and analyse such data


## The Command Line Interface (CLI)

[A text-based interface for giving instructions to a computer]{.highlighted} 

![](../img/terminal2.png){.absolute width=606px height=353px left=0px top=224.957px}

::: {.absolute width=437px height=272px left=612.932px top=270.821px style="font-size: 34px;"}
<ul>

<li>Allows to handle large datasets effectively</li>

<li>Gives us access to many bioinformatic tools</li>

<li>Easy to document and reproduce workflows</li>

</ul>
::: 


## High Performance Computing (HPC)

[A shared computer system that provides more memory, CPUs, space than a typical laptop]{.highlighted}

</br>

![](../img/crunchomics1.png)


## Real-life application

In this tutorial, you will analyse these reads to:

- Determine which strains are present in the mixed communities
- Quantify their relative abundances
- Asses whether the counts fit with your assumptions about individual interactions


## Workflow

</br>
</br>

```{mermaid}
flowchart LR
    classDef greenfill fill:#5B888C,stroke:#333,stroke-width:1,color:#fff;
    classDef dbfill fill:#E2F0F1,stroke:#333,stroke-width:1,color:#333;

    %% Workflow boxes
    A[Raw reads] -->|FASTQ| B[Quality control]
    B -->|FASTQ| C[Quality filtering]
    C -->|FASTQ| D[Mapping to <br> Reference DB]
    D -->|PAF| E[Count table]

    %% Reference database node
    DB[(16S <br> Reference  <br> Database)] -->|FASTA| D  

    %% Assign classes after nodes exist
    class A,B,C,D,E greenfill
    class DB dbfill

    %% Arrow styles (index starts at 0 for first arrow)
    linkStyle 0,1,2,3,4 stroke:#5B888C,stroke-width:2,color:#000, fill: none
```

</br>

::: {.absolute width=1038.16px height=160.597px left=-5.44323px top=490.062px style="font-size: 32px;"}

PAF: 

- Pairwise Alignment Format
- A table with information about how each read maps to the reference

:::


## The FastQ format 

[FASTQ format is a text-based format for storing a sequence and its corresponding quality scores]{.highlighted}

::: {.fragment}

![](../img/fastq_format_red.png){.absolute top=200px}

:::


## Phred scores

[The Phred score is a measure for base quality. The larger the Phred value, the better the quality of a base]{.highlighted}

::: {.fragment}

![](../img/phred1.png){.absolute top=200px}

:::


## Phred scores

[The Phred score is a measure for base quality. The larger the Phred value, the better the quality of a base]{.highlighted}

![](../img/phred2.png){.absolute top=200px}


## Bringing things together 

::: {.small}
```{bash}
#| echo: true
#| eval: false 

# Run Nanoplot on FASTQ files to assess quality
NanoPlot \                            # <1>
    --fastq data/barcode01.fastq \    # <2>
    -o QC_plots \                     # <2>
    -t 2 \                            # <2>
    --tsv_stats                       # <2>
```

1. Define what tool we want to use 
2. Use options to modify the tools behavior
:::

:::{.fragment}
![](../img/nanoplot.png){width=80%}
:::


## Practical part

You will have today and Monday to work on a practical and learn how to:

- Navigate the command line
- Work on an HPC
- Analyze sequencing data

{{< qrcode https://ndombrowski.github.io/MicEco2025// qr1 width=150 height=150 colorDark='#2d3e3f' >}}

[https://ndombrowski.github.io/MicEco2025/](https://ndombrowski.github.io/MicEco2025/){.small}