<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>prep_material</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="prep_material_files/libs/clipboard/clipboard.min.js"></script>
<script src="prep_material_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="prep_material_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="prep_material_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="prep_material_files/libs/quarto-html/popper.min.js"></script>
<script src="prep_material_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="prep_material_files/libs/quarto-html/anchor.min.js"></script>
<link href="prep_material_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="prep_material_files/libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="prep_material_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="prep_material_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="prep_material_files/libs/bootstrap/bootstrap-9e3ffae467580fdb927a41352e75a2e0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">




<section id="prep-course-material" class="level1">
<h1>Prep course material</h1>
<section id="list-of-nanopore-adaptors" class="level2">
<h2 class="anchored" data-anchor-id="list-of-nanopore-adaptors">List of nanopore adaptors</h2>
<p>For this tutorial I want to test fastplong as alternative to porechop followed by chopper. While fastplong should allow for automatic detection of adapters I also want to try and test it by providing a list of all possible adapter sequences.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /zfs/omics/projects/bioinformatics/scripts</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the porechop adapter script to get an input file for fastplong </span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://raw.githubusercontent.com/rrwick/Porechop/refs/heads/master/porechop/adapters.py</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Write script to extract all barcodes but output a fasta file of the full adapters</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ../files/nanopore </span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> parse_adapters.py</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>This script produces two files</p>
<ul>
<li>full_adapters.fasta (full adapters including all native and rapid barcodes)</li>
<li>barcode_only_adapters.fasta (just barcode sequences)</li>
</ul>
</section>
<section id="test-fastplong" class="level2">
<h2 class="anchored" data-anchor-id="test-fastplong">Test fastplong</h2>
<section id="installation" class="level3">
<h3 class="anchored" data-anchor-id="installation">Installation</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-p</span> /zfs/omics/projects/bioinformatics/software/miniconda3/envs/fastplong_0.4.1 <span class="at">-c</span> bioconda fastplong=0.4.1</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="default-settings" class="level3">
<h3 class="anchored" data-anchor-id="default-settings">Default settings</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> activate fastplong_0.4.1</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/fastplong/v1</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>20G <span class="at">--cpus-per-task</span> 2 fastplong <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">-i</span> data/barcode01_merged.fastq.gz <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mean_qual</span> 8 <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">-l</span> 1400 <span class="at">--length_limit</span> 1700 <span class="dt">\</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--thread</span> 2 <span class="dt">\</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> results/fastplong/v1/barcode01_filtered.fastq.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Defaults:</p>
<ul>
<li>-A, –disable_adapter_trimming adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled. If all these adapter options (start_adapter, end_adapter and adapter_fasta) are not specified, fastplong will try to detect the read start and read end adapters automatically. The detected adapter sequences may be a bit shorter or longer than the real ones. And there is a certain probability of misidentification, especially when most reads don’t have adapters (it won’t cause too bad result in this case).</li>
<li>-Q, –disable_quality_filtering quality filtering is enabled by default. If this option is specified, quality filtering is disabled</li>
<li>-q, –qualified_quality_phred the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified. (int [=15])</li>
<li>-W, –cut_window_size the window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4 (int [=4])</li>
<li>-M, –cut_mean_quality the mean quality requirement option shared by cut_front, cut_tail or cut_sliding. Range: 1~36 default: 20 (Q20) (int [=20])</li>
<li>-d, –distance_threshold threshold of sequence-adapter-distance/adapter-length (0.0 ~ 1.0), greater value means more adapters detected (double [=0.25])</li>
<li>–trimming_extension when an adapter is detected, extend the trimming to make cleaner trimming, default 10 means trimming 10 bases more (int [=10])</li>
</ul>
<p>Others:</p>
<ul>
<li>-f, –trim_front trimming how many bases in front for read, default is 0 (int [=0])</li>
<li>-5, –cut_front move a sliding window from front (5’) to tail, drop the bases in the window if its mean quality &lt; threshold, stop otherwise.</li>
<li>-3, –cut_tail move a sliding window from tail (3’) to front, drop the bases in the window if its mean quality &lt; threshold, stop otherwise.</li>
<li>-m, –mean_qual if one read’s mean_qual quality score &lt; mean_qual, then this read is discarded. Default 0 means no requirement (int [=0])</li>
</ul>
<p>Results (1 seconds):</p>
<pre><code>Trying to detect adapter sequence at read start
Detected: GCCTACCGTGAC{AAGAAAGTTGTCGGTGTCTTTGTG}  {barcode1_start_sequence}
Trying to detect adapter sequence at read end   
Detected: {CACAAAGACACCGACAACTTTCTT}GTC            {barcode1_end_sequence}    

Before filtering:
total reads: 17121
total bases: 24089162
Q20 bases: 13250840(55.0075%)
Q30 bases: 4013925(16.6628%)

After filtering:
total reads: 8441
total bases: 12359581
Q20 bases: 7827054(63.3278%)
Q30 bases: 2512370(20.3273%)

Filtering result:
reads passed filter: 8441
reads failed due to low quality: 4136
reads failed due to too many N: 0
reads failed due to too short: 4569
reads failed due to too long: 15
reads with adapter trimmed: 16548
bases trimmed due to adapters: 1808399

JSON report: fastplong.json
HTML report: fastplong.html</code></pre>
<p>Before: <span class="citation" data-cites="a72b920c-8266-4865-84a9-50690efe5943">@a72b920c-8266-4865-84a9-50690efe5943</span> runid=b1ae45334215a9c23ceb122ef2b45224cee1eeab read=26381 ch=122 start_time=2022-03-10T20:32:29Z flow_cell_id=AHJ236 protocol_group_id=matteo2 sample_id=mt2 barcode=barcode01 barcode_alias=barcode01 TTGTACTTCGTTCAGTTCGGGCATTGGGTGTTGCTCAGTTC[ATCGCCTACCGTGAC]{AAGAAAGTTGTCGGTGTCTTTGTG}AGAGTTTGAT|start_trimming|CATGGCTCAGGATGAACGCTGGCGGCGTACCTAACACATGCAAGTCGAACGGGTACCTTTGACAGAGATTTCGGTCAATTGATAAGGATACTTAGTGGCGGACGGGTGAGGTAACGCGTGAGCAATCTGCCTTTCAGTGGGGGAATAACAGTCGGAAACGATTGCTAATACCGCATAATGTATCGGGATCGCATGGTTCTGATACCAAAGATTTATCGCTGAAAGATGAGCTCGCGTCTGATTAGCTAGTTGGTAAGGTAATGGCTTGCAAAGCGACGATCAGTAGCCGGACTGAAGGAGTTGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACAGGAGGCAGCAGTGGGAATATTGGGCAATGGGCATAAGCCTGACCCAGCAACGCCGCGTGAAGGAAAAGGCTTTCGGGTTGTAAACTTCTTTTATCAGGGACGAAAAAAATGACGGTACCTGAAGAATAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGGACTGCAAGTCAGATGTGAAATCTATGGGCTCAACCCATAGCGTGCATTTGAAACTGTAGTTCTTGAGTGATGGAGAGGCAAACGGAATTCCGTGTGTAGCGGTGAAATGCGTAGATATACGGAGGAACACCAGTGGCGAAGGCGGTTTGCTGGACATTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGAGTTAGATACCCTGGTAATTTTTGCTGTAAACGATGGATACTAGGTGTGGGGGGACTGACCCCTTCCGTGCCGCAGTTAACACAATAAGTATCCCACCTGGGGAGTACGATCGCAAGGTCGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGGTATGTGGATTAATTCGAAGCAACGCGAAGAACCTTACCAGGACTTGACATCCTATAACGAAGTAGAGATACATTAGGTGCCCTTCGGGGAAAGTAGAACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAGAATTCTATGAGCTGCTTATATTAGTTGCTACGCAAGAGCACTCTAGCGAGATACCGTTGATAAAACAGGAGGAAGATTGGGGACGACGTCAAATCATCATGCCTCTTATGTCCTGGGCCTCACGTACTACAATGGCGATTAACAGAGGGAAGCAAGACCGCGAGGTGGAGCAAATCCTAAATCGTCTCAGTTCAGATTGCAGGCTGAAACCCGCCTGTATGAAGTCGGAATCGCTAGTAATCGCGGATCCAGCATGCCGCGGTGAATACGTTCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTCAGGAACACCGAAGTCCGTAGCTTAACCGCAAGGAGCGCGGCCGAAGGTGGGTTTGATAATTGGGGGTGAAGTCGTAAC|end trimming|AAGGTAACCG{CACAAAGACACCGACAACTTTCTT}G</p>
<p>After: <span class="citation" data-cites="a72b920c-8266-4865-84a9-50690efe5943">@a72b920c-8266-4865-84a9-50690efe5943</span> runid=b1ae45334215a9c23ceb122ef2b45224cee1eeab read=26381 ch=122 start_time=2022-03-10T20:32:29Z flow_cell_id=AHJ236 protocol_group_id=matteo2 sample_id=mt2 barcode=barcode01 barcode_alias=barcode01 CATGGCTCAGGATGAACGCTGGCGGCGTACCTAACACATGCAAGTCGAACGGGTACCTTTGACAGAGATTTCGGTCAATTGATAAGGATACTTAGTGGCGGACGGGTGAGGTAACGCGTGAGCAATCTGCCTTTCAGTGGGGGAATAACAGTCGGAAACGATTGCTAATACCGCATAATGTATCGGGATCGCATGGTTCTGATACCAAAGATTTATCGCTGAAAGATGAGCTCGCGTCTGATTAGCTAGTTGGTAAGGTAATGGCTTGCAAAGCGACGATCAGTAGCCGGACTGAAGGAGTTGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACAGGAGGCAGCAGTGGGAATATTGGGCAATGGGCATAAGCCTGACCCAGCAACGCCGCGTGAAGGAAAAGGCTTTCGGGTTGTAAACTTCTTTTATCAGGGACGAAAAAAATGACGGTACCTGAAGAATAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGGACTGCAAGTCAGATGTGAAATCTATGGGCTCAACCCATAGCGTGCATTTGAAACTGTAGTTCTTGAGTGATGGAGAGGCAAACGGAATTCCGTGTGTAGCGGTGAAATGCGTAGATATACGGAGGAACACCAGTGGCGAAGGCGGTTTGCTGGACATTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGAGTTAGATACCCTGGTAATTTTTGCTGTAAACGATGGATACTAGGTGTGGGGGGACTGACCCCTTCCGTGCCGCAGTTAACACAATAAGTATCCCACCTGGGGAGTACGATCGCAAGGTCGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGGTATGTGGATTAATTCGAAGCAACGCGAAGAACCTTACCAGGACTTGACATCCTATAACGAAGTAGAGATACATTAGGTGCCCTTCGGGGAAAGTAGAACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAGAATTCTATGAGCTGCTTATATTAGTTGCTACGCAAGAGCACTCTAGCGAGATACCGTTGATAAAACAGGAGGAAGATTGGGGACGACGTCAAATCATCATGCCTCTTATGTCCTGGGCCTCACGTACTACAATGGCGATTAACAGAGGGAAGCAAGACCGCGAGGTGGAGCAAATCCTAAATCGTCTCAGTTCAGATTGCAGGCTGAAACCCGCCTGTATGAAGTCGGAATCGCTAGTAATCGCGGATCCAGCATGCCGCGGTGAATACGTTCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTCAGGAACACCGAAGTCCGTAGCTTAACCGCAAGGAGCGCGGCCGAAGGTGGGTTTGATAATTGGGGGTGAAGTCGTAAC</p>
<p><a href="https://nanoporetech.com/document/data-analysis#barcoding-options">Barcode design</a></p>
<p>The Oxford Nanopore Technologies barcoding kits can place barcodes at the beginning and for some kits, also at the end of the strands for multiplexing several different samples in one sequencing experiment. The barcodes will reside in a kit-specific context sequence, and different kits will have different lengths of sequence before and after the barcode. However, the sequences of the barcodes themselves are identical, regardless of kit.</p>
<p>A complete barcode arrangement comprises three sections:</p>
<pre><code>- The upstream flanking region, which comes between the barcode and the sequencing adapter.
- The barcode sequence.
- The downstream flanking region, which comes between the barcode and the sample sequence.</code></pre>
</section>
<section id="using-a-adapter-file" class="level3">
<h3 class="anchored" data-anchor-id="using-a-adapter-file">Using a adapter file</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> results/fastplong/v2 </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>20G <span class="at">--cpus-per-task</span> 10 fastplong <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-i</span> data/barcode01_merged.fastq.gz <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mean_qual</span> 8 <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">-l</span> 1400 <span class="at">--length_limit</span> 1700 <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">-a</span> /zfs/omics/projects/bioinformatics/files/nanopore/all_sequences.fasta <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--thread</span> 10 <span class="dt">\</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> results/fastplong/v2/barcode01_filtered.fastq.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Results (94s seconds):</p>
<pre><code>Trying to detect adapter sequence at read start
Detected: GCCTACCGTGACAAGAAAGTTGTCGGTGTCTTTGTG
Trying to detect adapter sequence at read end
Detected: CACAAAGACACCGACAACTTTCTTGTC
&lt;longer wait time here, likely since its now checking the other sequences&gt; 

Before filtering:
total reads: 17121
total bases: 24089162
Q20 bases: 13250840(55.0075%)
Q30 bases: 4013925(16.6628%)

After filtering:
total reads: 7392
total bases: 10801616
Q20 bases: 6838544(63.3104%)
Q30 bases: 2192504(20.2979%)

Filtering result:
reads passed filter: 7392
reads failed due to low quality: 4092
reads failed due to too many N: 0
reads failed due to too short: 5660
reads failed due to too long: 15
reads with adapter trimmed: 16921
bases trimmed due to adapters: 2182059

JSON report: fastplong.json
HTML report: fastplong.html</code></pre>
<p>Before: <span class="citation" data-cites="a72b920c-8266-4865-84a9-50690efe5943">@a72b920c-8266-4865-84a9-50690efe5943</span> runid=b1ae45334215a9c23ceb122ef2b45224cee1eeab read=26381 ch=122 start_time=2022-03-10T20:32:29Z flow_cell_id=AHJ236 protocol_group_id=matteo2 sample_id=mt2 barcode=barcode01 barcode_alias=barcode01 TTGTACTTCGTTCAGTTCGGGCATTGGGTGTTGCTCAGTTC[ATCGCCTACCGTGAC]{AAGAAAGTTGTCGGTGTCTTTGTG}AGAGTTTGAT|start_trimming|CATGGCTCAGGATGAACGCTGGCGGCGTACCTAACACATGCAAGTCGAACGGGTACCTTTGACAGAGATTTCGGTCAATTGATAAGGATACTTAGTGGCGGACGGGTGAGGTAACGCGTGAGCAATCTGCCTTTCAGTGGGGGAATAACAGTCGGAAACGATTGCTAATACCGCATAATGTATCGGGATCGCATGGTTCTGATACCAAAGATTTATCGCTGAAAGATGAGCTCGCGTCTGATTAGCTAGTTGGTAAGGTAATGGCTTGCAAAGCGACGATCAGTAGCCGGACTGAAGGAGTTGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACAGGAGGCAGCAGTGGGAATATTGGGCAATGGGCATAAGCCTGACCCAGCAACGCCGCGTGAAGGAAAAGGCTTTCGGGTTGTAAACTTCTTTTATCAGGGACGAAAAAAATGACGGTACCTGAAGAATAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGGACTGCAAGTCAGATGTGAAATCTATGGGCTCAACCCATAGCGTGCATTTGAAACTGTAGTTCTTGAGTGATGGAGAGGCAAACGGAATTCCGTGTGTAGCGGTGAAATGCGTAGATATACGGAGGAACACCAGTGGCGAAGGCGGTTTGCTGGACATTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGAGTTAGATACCCTGGTAATTTTTGCTGTAAACGATGGATACTAGGTGTGGGGGGACTGACCCCTTCCGTGCCGCAGTTAACACAATAAGTATCCCACCTGGGGAGTACGATCGCAAGGTCGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGGTATGTGGATTAATTCGAAGCAACGCGAAGAACCTTACCAGGACTTGACATCCTATAACGAAGTAGAGATACATTAGGTGCCCTTCGGGGAAAGTAGAACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAGAATTCTATGAGCTGCTTATATTAGTTGCTACGCAAGAGCACTCTAGCGAGATACCGTTGATAAAACAGGAGGAAGATTGGGGACGACGTCAAATCATCATGCCTCTTATGTCCTGGGCCTCACGTACTACAATGGCGATTAACAGAGGGAAGCAAGACCGCGAGGTGGAGCAAATCCTAAATCGTCTCAGTTCAGATTGCAGGCTGAAACCCGCCTGTATGAAGTCGGAATCGCTAGTAATCGCGGATCCAGCATGCCGCGGTGAATACGTTCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTCAGGAACACCGAAGTCCGTAGCTTAACCGCAAGGAGCGCGGCCGAAGGTGGGTTTGATAATTGGGGGTGAAGTCGTAAC|end trimming|AAGGTAACCG{CACAAAGACACCGACAACTTTCTT}G</p>
<p><span class="citation" data-cites="a72b920c-8266-4865-84a9-50690efe5943">@a72b920c-8266-4865-84a9-50690efe5943</span> runid=b1ae45334215a9c23ceb122ef2b45224cee1eeab read=26381 ch=122 start_time=2022-03-10T20:32:29Z flow_cell_id=AHJ236 protocol_group_id=matteo2 sample_id=mt2 barcode=barcode01 barcode_alias=barcode01 CATGGCTCAGGATGAACGCTGGCGGCGTACCTAACACATGCAAGTCGAACGGGTACCTTTGACAGAGATTTCGGTCAATTGATAAGGATACTTAGTGGCGGACGGGTGAGGTAACGCGTGAGCAATCTGCCTTTCAGTGGGGGAATAACAGTCGGAAACGATTGCTAATACCGCATAATGTATCGGGATCGCATGGTTCTGATACCAAAGATTTATCGCTGAAAGATGAGCTCGCGTCTGATTAGCTAGTTGGTAAGGTAATGGCTTGCAAAGCGACGATCAGTAGCCGGACTGAAGGAGTTGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACAGGAGGCAGCAGTGGGAATATTGGGCAATGGGCATAAGCCTGACCCAGCAACGCCGCGTGAAGGAAAAGGCTTTCGGGTTGTAAACTTCTTTTATCAGGGACGAAAAAAATGACGGTACCTGAAGAATAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGGACTGCAAGTCAGATGTGAAATCTATGGGCTCAACCCATAGCGTGCATTTGAAACTGTAGTTCTTGAGTGATGGAGAGGCAAACGGAATTCCGTGTGTAGCGGTGAAATGCGTAGATATACGGAGGAACACCAGTGGCGAAGGCGGTTTGCTGGACATTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGAGTTAGATACCCTGGTAATTTTTGCTGTAAACGATGGATACTAGGTGTGGGGGGACTGACCCCTTCCGTGCCGCAGTTAACACAATAAGTATCCCACCTGGGGAGTACGATCGCAAGGTCGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGGTATGTGGATTAATTCGAAGCAACGCGAAGAACCTTACCAGGACTTGACATCCTATAACGAAGTAGAGATACATTAGGTGCCCTTCGGGGAAAGTAGAACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAGAATTCTATGAGCTGCTTATATTAGTTGCTACGCAAGAGCACTCTAGCGAGATACCGTTGATAAAACAGGAGGAAGATTGGGGACGACGTCAAATCATCATGCCTCTTATGTCCTGGGCCTCACGTACTACAATGGCGATTAACAGAGGGAAGCAAGACCGCGAGGTGGAGCAAATCCTAAATCGTCTCAGTTCAGATTGCAGGCTGAAACCCGCCTGTATGAAGTCGGAATCGCTAGTAATCGCGGATCCAGCATGCCGCGGTGAATACGTTCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTCAGGAACACCGAAGTCCGTAGCTTAACCGCAAGGAGCGCGGCCGAAGGTGGGTTTGATAATTGGGGGTGAAGTCGTAAC</p>
</section>
<section id="using-a-shortened-adapter-file" class="level3">
<h3 class="anchored" data-anchor-id="using-a-shortened-adapter-file">Using a shortened adapter file</h3>
<p>The analysis above seems more accurate but runs quite a bit with 356 adapter sequences. Since this contains individual barcodes but also full sequences, I shortened this file to only include individual sequences, since filtlong will remove any part before and after the forward and reverse barcodes (211).</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> results/fastplong/v3</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>20G <span class="at">--cpus-per-task</span> 10 fastplong <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-i</span> data/barcode01_merged.fastq.gz <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mean_qual</span> 8 <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">-l</span> 1400 <span class="at">--length_limit</span> 1700 <span class="dt">\</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">-a</span> /zfs/omics/projects/bioinformatics/files/nanopore/all_sequences_unique.fasta <span class="dt">\</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--thread</span> 10 <span class="dt">\</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> results/fastplong/v3/barcode01_filtered.fastq.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Before: <span class="citation" data-cites="a72b920c-8266-4865-84a9-50690efe5943">@a72b920c-8266-4865-84a9-50690efe5943</span> runid=b1ae45334215a9c23ceb122ef2b45224cee1eeab read=26381 ch=122 start_time=2022-03-10T20:32:29Z flow_cell_id=AHJ236 protocol_group_id=matteo2 sample_id=mt2 barcode=barcode01 barcode_alias=barcode01 TTGTACTTCGTTCAGTTCGGGCATTGGGTGTTGCTCAGTTC[ATCGCCTACCGTGAC]{AAGAAAGTTGTCGGTGTCTTTGTG}AGAGTTTGAT|start_trimming|CATGGCTCAGGATGAACGCTGGCGGCGTACCTAACACATGCAAGTCGAACGGGTACCTTTGACAGAGATTTCGGTCAATTGATAAGGATACTTAGTGGCGGACGGGTGAGGTAACGCGTGAGCAATCTGCCTTTCAGTGGGGGAATAACAGTCGGAAACGATTGCTAATACCGCATAATGTATCGGGATCGCATGGTTCTGATACCAAAGATTTATCGCTGAAAGATGAGCTCGCGTCTGATTAGCTAGTTGGTAAGGTAATGGCTTGCAAAGCGACGATCAGTAGCCGGACTGAAGGAGTTGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACAGGAGGCAGCAGTGGGAATATTGGGCAATGGGCATAAGCCTGACCCAGCAACGCCGCGTGAAGGAAAAGGCTTTCGGGTTGTAAACTTCTTTTATCAGGGACGAAAAAAATGACGGTACCTGAAGAATAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGGACTGCAAGTCAGATGTGAAATCTATGGGCTCAACCCATAGCGTGCATTTGAAACTGTAGTTCTTGAGTGATGGAGAGGCAAACGGAATTCCGTGTGTAGCGGTGAAATGCGTAGATATACGGAGGAACACCAGTGGCGAAGGCGGTTTGCTGGACATTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGAGTTAGATACCCTGGTAATTTTTGCTGTAAACGATGGATACTAGGTGTGGGGGGACTGACCCCTTCCGTGCCGCAGTTAACACAATAAGTATCCCACCTGGGGAGTACGATCGCAAGGTCGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGGTATGTGGATTAATTCGAAGCAACGCGAAGAACCTTACCAGGACTTGACATCCTATAACGAAGTAGAGATACATTAGGTGCCCTTCGGGGAAAGTAGAACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAGAATTCTATGAGCTGCTTATATTAGTTGCTACGCAAGAGCACTCTAGCGAGATACCGTTGATAAAACAGGAGGAAGATTGGGGACGACGTCAAATCATCATGCCTCTTATGTCCTGGGCCTCACGTACTACAATGGCGATTAACAGAGGGAAGCAAGACCGCGAGGTGGAGCAAATCCTAAATCGTCTCAGTTCAGATTGCAGGCTGAAACCCGCCTGTATGAAGTCGGAATCGCTAGTAATCGCGGATCCAGCATGCCGCGGTGAATACGTTCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTCAGGAACACCGAAGTCCGTAGCTTAACCGCAAGGAGCGCGGCCGAAGGTGGGTTTGATAATTGGGGGTGAAGTCGTAAC|end trimming|AAGGTAACCG{CACAAAGACACCGACAACTTTCTT}G</p>
<p>After filtlong: <span class="citation" data-cites="a72b920c-8266-4865-84a9-50690efe5943">@a72b920c-8266-4865-84a9-50690efe5943</span> runid=b1ae45334215a9c23ceb122ef2b45224cee1eeab read=26381 ch=122 start_time=2022-03-10T20:32:29Z flow_cell_id=AHJ236 protocol_group_id=matteo2 sample_id=mt2 barcode=barcode01 barcode_alias=barcode01 CATGGCTCAGGATGAACGCTGGCGGCGTACCTAACACATGCAAGTCGAACGGGTACCTTTGACAGAGATTTCGGTCAATTGATAAGGATACTTAGTGGCGGACGGGTGAGGTAACGCGTGAGCAATCTGCCTTTCAGTGGGGGAATAACAGTCGGAAACGATTGCTAATACCGCATAATGTATCGGGATCGCATGGTTCTGATACCAAAGATTTATCGCTGAAAGATGAGCTCGCGTCTGATTAGCTAGTTGGTAAGGTAATGGCTTGCAAAGCGACGATCAGTAGCCGGACTGAAGGAGTTGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACAGGAGGCAGCAGTGGGAATATTGGGCAATGGGCATAAGCCTGACCCAGCAACGCCGCGTGAAGGAAAAGGCTTTCGGGTTGTAAACTTCTTTTATCAGGGACGAAAAAAATGACGGTACCTGAAGAATAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGGACTGCAAGTCAGATGTGAAATCTATGGGCTCAACCCATAGCGTGCATTTGAAACTGTAGTTCTTGAGTGATGGAGAGGCAAACGGAATTCCGTGTGTAGCGGTGAAATGCGTAGATATACGGAGGAACACCAGTGGCGAAGGCGGTTTGCTGGACATTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGAGTTAGATACCCTGGTAATTTTTGCTGTAAACGATGGATACTAGGTGTGGGGGGACTGACCCCTTCCGTGCCGCAGTTAACACAATAAGTATCCCACCTGGGGAGTACGATCGCAAGGTCGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGGTATGTGGATTAATTCGAAGCAACGCGAAGAACCTTACCAGGACTTGACATCCTATAACGAAGTAGAGATACATTAGGTGCCCTTCGGGGAAAGTAGAACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAGAATTCTATGAGCTGCTTATATTAGTTGCTACGCAAGAGCACTCTAGCGAGATACCGTTGATAAAACAGGAGGAAGATTGGGGACGACGTCAAATCATCATGCCTCTTATGTCCTGGGCCTCACGTACTACAATGGCGATTAACAGAGGGAAGCAAGACCGCGAGGTGGAGCAAATCCTAAATCGTCTCAGTTCAGATTGCAGGCTGAAACCCGCCTGTATGAAGTCGGAATCGCTAGTAATCGCGGATCCAGCATGCCGCGGTGAATACGTTCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTCAGGAACACCGAAGTCCGTAGCTTAACCGCAAGGAGCGCGGCCGAAGGTGGGTTTGATAATTGGGGGTGAAGTCGTAAC</p>
<p>Results (56 seconds):</p>
<pre><code>Trying to detect adapter sequence at read start
Detected: GCCTACCGTGACAAGAAAGTTGTCGGTGTCTTTGTG
Trying to detect adapter sequence at read end
Detected: CACAAAGACACCGACAACTTTCTTGTC

Before filtering:
total reads: 17121
total bases: 24089162
Q20 bases: 13250840(55.0075%)
Q30 bases: 4013925(16.6628%)

After filtering:
total reads: 7393
total bases: 10803314
Q20 bases: 6839277(63.3072%)
Q30 bases: 2192695(20.2965%)

Filtering result:
reads passed filter: 7393
reads failed due to low quality: 4090
reads failed due to too many N: 0
reads failed due to too short: 5661
reads failed due to too long: 15
reads with adapter trimmed: 16,919
bases trimmed due to adapters: 2,180,983</code></pre>
<table class="caption-top table">
<colgroup>
<col style="width: 28%">
<col style="width: 13%">
<col style="width: 25%">
<col style="width: 33%">
</colgroup>
<thead>
<tr class="header">
<th>Metric</th>
<th>Default (v1)</th>
<th>With adapter FASTA (v2)</th>
<th>With shorter adapter fasta (v3)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Reads passed filter</td>
<td>8441</td>
<td>7392</td>
<td>7393</td>
</tr>
<tr class="even">
<td>Reads failed short</td>
<td>4569</td>
<td>5660</td>
<td>5661</td>
</tr>
<tr class="odd">
<td>Reads with adapter trimmed</td>
<td>16548</td>
<td>16921</td>
<td>16919</td>
</tr>
<tr class="even">
<td>Bases trimmed</td>
<td>1,808,399</td>
<td>2,182,059</td>
<td>2,180,983</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="compare-old-setup-with-fastplong" class="level2">
<h2 class="anchored" data-anchor-id="compare-old-setup-with-fastplong">Compare old setup with fastplong</h2>
<section id="porechop" class="level3">
<h3 class="anchored" data-anchor-id="porechop">Porechop</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate porechop_0.2.4</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> results/old/porechop/</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="ex">porechop</span> <span class="at">--input</span> data/barcode01_merged.fastq.gz <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--output</span> results/old/porechop/barcode01.fastq.gz <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--threads</span> 2 <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--discard_middle</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<table class="caption-top table">
<colgroup>
<col style="width: 27%">
<col style="width: 5%">
<col style="width: 20%">
<col style="width: 46%">
</colgroup>
<thead>
<tr class="header">
<th>Metric</th>
<th>Filtlong</th>
<th>Porechop</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Total reads</td>
<td>17,121</td>
<td>17,121</td>
<td>Same dataset</td>
</tr>
<tr class="even">
<td>Reads with adapters trimmed (start or end)</td>
<td>16,919</td>
<td>Start: 16,530 / End: 11,881</td>
<td>Filtlong trims more reads overall; Porechop distinguishes start vs end</td>
</tr>
<tr class="odd">
<td>Bases trimmed due to adapters</td>
<td>2,180,983</td>
<td>Start: 1,342,645 / End: 331,934</td>
<td>Filtlong trims more bases overall</td>
</tr>
<tr class="even">
<td>Reads discarded due to middle adapters</td>
<td>N/A</td>
<td>38</td>
<td>Filtlong does not report middle adapter trimming by default</td>
</tr>
</tbody>
</table>
<p>More bp removed with fastplong might be due to <code>-  --trimming_extension           when an adapter is detected, extend the trimming to make cleaner trimming, default 10 means trimming 10 bases more (int [=10])</code> For an exact comparison let’s remove this:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> activate fastplong_0.4.1</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> results/fastplong/v4</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>20G <span class="at">--cpus-per-task</span> 10 fastplong <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">-i</span> data/barcode01_merged.fastq.gz <span class="dt">\</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">-Q</span> <span class="at">-L</span> <span class="at">--trimming_extension</span> 0 <span class="dt">\</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">-a</span> /zfs/omics/projects/bioinformatics/files/nanopore/all_sequences_unique.fasta <span class="dt">\</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--thread</span> 10 <span class="dt">\</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> results/fastplong/v4/barcode01_filtered.fastq.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<table class="caption-top table">
<colgroup>
<col style="width: 27%">
<col style="width: 5%">
<col style="width: 20%">
<col style="width: 46%">
</colgroup>
<thead>
<tr class="header">
<th>Metric</th>
<th>Filtlong</th>
<th>Porechop</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Total reads</td>
<td>17,121</td>
<td>17,121</td>
<td>Same dataset</td>
</tr>
<tr class="even">
<td>Reads with adapters trimmed (start or end)</td>
<td>16,919</td>
<td>Start: 16,530 / End: 11,881</td>
<td>Filtlong trims more reads overall; Porechop distinguishes start vs end</td>
</tr>
<tr class="odd">
<td>Bases trimmed due to adapters</td>
<td>1,893,740</td>
<td>Start: 1,342,645 / End: 331,934</td>
<td>Filtlong trims more bases overall</td>
</tr>
<tr class="even">
<td>Reads discarded due to middle adapters</td>
<td>N/A</td>
<td>38</td>
<td>Filtlong does not report middle adapter trimming by default</td>
</tr>
</tbody>
</table>
<p>Filtlong consistently trims slightly more reads at start/end than Porechop, even when –trimming_extension is set to 0.</p>
<p>Before: <span class="citation" data-cites="a72b920c-8266-4865-84a9-50690efe5943">@a72b920c-8266-4865-84a9-50690efe5943</span> runid=b1ae45334215a9c23ceb122ef2b45224cee1eeab read=26381 ch=122 start_time=2022-03-10T20:32:29Z flow_cell_id=AHJ236 protocol_group_id=matteo2 sample_id=mt2 barcode=barcode01 barcode_alias=barcode01 TTGTACTTCGTTCAGTTCGGGCATTGGGTGTTGCTCAGTTC[ATCGCCTACCGTGAC]{AAGAAAGTTGTCGGTGTCTTTGTG}AG|start_trimming|AGTTTGATCATGGCTCAGGATGAACGCTGGCGGCGTACCTAACACATGCAAGTCGAACGGGTACCTTTGACAGAGATTTCGGTCAATTGATAAGGATACTTAGTGGCGGACGGGTGAGGTAACGCGTGAGCAATCTGCCTTTCAGTGGGGGAATAACAGTCGGAAACGATTGCTAATACCGCATAATGTATCGGGATCGCATGGTTCTGATACCAAAGATTTATCGCTGAAAGATGAGCTCGCGTCTGATTAGCTAGTTGGTAAGGTAATGGCTTGCAAAGCGACGATCAGTAGCCGGACTGAAGGAGTTGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACAGGAGGCAGCAGTGGGAATATTGGGCAATGGGCATAAGCCTGACCCAGCAACGCCGCGTGAAGGAAAAGGCTTTCGGGTTGTAAACTTCTTTTATCAGGGACGAAAAAAATGACGGTACCTGAAGAATAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGGACTGCAAGTCAGATGTGAAATCTATGGGCTCAACCCATAGCGTGCATTTGAAACTGTAGTTCTTGAGTGATGGAGAGGCAAACGGAATTCCGTGTGTAGCGGTGAAATGCGTAGATATACGGAGGAACACCAGTGGCGAAGGCGGTTTGCTGGACATTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGAGTTAGATACCCTGGTAATTTTTGCTGTAAACGATGGATACTAGGTGTGGGGGGACTGACCCCTTCCGTGCCGCAGTTAACACAATAAGTATCCCACCTGGGGAGTACGATCGCAAGGTCGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGGTATGTGGATTAATTCGAAGCAACGCGAAGAACCTTACCAGGACTTGACATCCTATAACGAAGTAGAGATACATTAGGTGCCCTTCGGGGAAAGTAGAACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAGAATTCTATGAGCTGCTTATATTAGTTGCTACGCAAGAGCACTCTAGCGAGATACCGTTGATAAAACAGGAGGAAGATTGGGGACGACGTCAAATCATCATGCCTCTTATGTCCTGGGCCTCACGTACTACAATGGCGATTAACAGAGGGAAGCAAGACCGCGAGGTGGAGCAAATCCTAAATCGTCTCAGTTCAGATTGCAGGCTGAAACCCGCCTGTATGAAGTCGGAATCGCTAGTAATCGCGGATCCAGCATGCCGCGGTGAATACGTTCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTCAGGAACACCGAAGTCCGTAGCTTAACCGCAAGGAGCGCGGCCGAAGGTGGGTTTGATAATTGGGGGTGAAGTCGTAAC|end trimming|AAGGTAACCG{CACAAAGACACCGACAACTTTCTT}G</p>
<p>After porechop: <span class="citation" data-cites="a72b920c-8266-4865-84a9-50690efe5943">@a72b920c-8266-4865-84a9-50690efe5943</span> runid=b1ae45334215a9c23ceb122ef2b45224cee1eeab read=26381 ch=122 start_time=2022-03-10T20:32:29Z flow_cell_id=AHJ236 protocol_group_id=matteo2 sample_id=mt2 barcode=barcode01 barcode_alias=barcode01 AGTTTGATCATGGCTCAGGATGAACGCTGGCGGCGTACCTAACACATGCAAGTCGAACGGGTACCTTTGACAGAGATTTCGGTCAATTGATAAGGATACTTAGTGGCGGACGGGTGAGGTAACGCGTGAGCAATCTGCCTTTCAGTGGGGGAATAACAGTCGGAAACGATTGCTAATACCGCATAATGTATCGGGATCGCATGGTTCTGATACCAAAGATTTATCGCTGAAAGATGAGCTCGCGTCTGATTAGCTAGTTGGTAAGGTAATGGCTTGCAAAGCGACGATCAGTAGCCGGACTGAAGGAGTTGAACGGCCACATTGGGACTGAGACACGGCCCAGACTCCTACAGGAGGCAGCAGTGGGAATATTGGGCAATGGGCATAAGCCTGACCCAGCAACGCCGCGTGAAGGAAAAGGCTTTCGGGTTGTAAACTTCTTTTATCAGGGACGAAAAAAATGACGGTACCTGAAGAATAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGGACTGCAAGTCAGATGTGAAATCTATGGGCTCAACCCATAGCGTGCATTTGAAACTGTAGTTCTTGAGTGATGGAGAGGCAAACGGAATTCCGTGTGTAGCGGTGAAATGCGTAGATATACGGAGGAACACCAGTGGCGAAGGCGGTTTGCTGGACATTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGAGTTAGATACCCTGGTAATTTTTGCTGTAAACGATGGATACTAGGTGTGGGGGGACTGACCCCTTCCGTGCCGCAGTTAACACAATAAGTATCCCACCTGGGGAGTACGATCGCAAGGTCGAAACTCAAAGGAATTGACGGGGGCCCGCACAAGCGGTGGGTATGTGGATTAATTCGAAGCAACGCGAAGAACCTTACCAGGACTTGACATCCTATAACGAAGTAGAGATACATTAGGTGCCCTTCGGGGAAAGTAGAACAGGTGGTGCATGGTTGTCGTCAGCTCGTGTCGTGAGATGTTAGGTTAGAATTCTATGAGCTGCTTATATTAGTTGCTACGCAAGAGCACTCTAGCGAGATACCGTTGATAAAACAGGAGGAAGATTGGGGACGACGTCAAATCATCATGCCTCTTATGTCCTGGGCCTCACGTACTACAATGGCGATTAACAGAGGGAAGCAAGACCGCGAGGTGGAGCAAATCCTAAATCGTCTCAGTTCAGATTGCAGGCTGAAACCCGCCTGTATGAAGTCGGAATCGCTAGTAATCGCGGATCCAGCATGCCGCGGTGAATACGTTCGGGCCTTGTACACACCGCCCGTCACACCACGAGAGTCAGGAACACCGAAGTCCGTAGCTTAACCGCAAGGAGCGCGGCCGAAGGTGGGTTTGATAATTGGGGGTGAAGTCGTAACAAGGTAAC</p>
</section>
<section id="summary" class="level3">
<h3 class="anchored" data-anchor-id="summary">Summary</h3>
<ul>
<li>Default Filtlong behavior
<ul>
<li>fastplong detects the correct barcodes at the start and end of reads without providing an adapters file.</li>
<li>Trimming effectively removes any sequence before or after the barcode.</li>
</ul></li>
<li>Providing adapter sequences
<ul>
<li>Slightly increases the number of bases and reads trimmed.</li>
<li>Can capture partial or unusual adapter sequences that default detection might miss.</li>
<li>Runtime is longer due to the larger adapter library and additional sequence comparisons.</li>
</ul></li>
<li>Comparison with Porechop
<ul>
<li>Both tools trim a similar number of reads and bases.</li>
<li>Filtlong appears to remove slightly more sequence. Filtlong removes everything before the barcode at the start and trims slightly more at the end (adapter + extra bases possibly due to mismatch allowance and <code>--trimming_extension 10</code>).</li>
<li>Porechop uses alignment scoring to handle errors, but exact sequence searching shows why raw reads rarely contain perfect adapter sequences.</li>
</ul></li>
<li>Practical considerations
<ul>
<li>Porechop is no longer maintained.</li>
<li>Subsequent workflows often require chopper for additional quality trimming.</li>
<li>fastplong provides a alternative</li>
</ul></li>
</ul>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>