<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>test_clustering_v1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="test_clustering_v1_files/libs/clipboard/clipboard.min.js"></script>
<script src="test_clustering_v1_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="test_clustering_v1_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="test_clustering_v1_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="test_clustering_v1_files/libs/quarto-html/popper.min.js"></script>
<script src="test_clustering_v1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="test_clustering_v1_files/libs/quarto-html/anchor.min.js"></script>
<link href="test_clustering_v1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="test_clustering_v1_files/libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="test_clustering_v1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="test_clustering_v1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="test_clustering_v1_files/libs/bootstrap/bootstrap-9e3ffae467580fdb927a41352e75a2e0.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">




<section id="test-clustering" class="level1">
<h1>Test clustering</h1>
<p>To develop the workflow I reveived 15 barcodes of fastq data from Isabel Rathmann, each barcode is nanopore reads from a single microbial species. I will combine all these reads into one to see if we ideally can reconstruct up to 15 consensus sequences. I first will integrate the data into the current workflow for quality assessment and cleaning and then start testing tools like NGSpeciesID for clustering.</p>
<section id="organize-test-files" class="level2">
<h2 class="anchored" data-anchor-id="organize-test-files">Organize test files</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/ndombro/personal/data_analysis/</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Path to your base folder</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="va">base_folder</span><span class="op">=</span><span class="st">"/scratch/transfer_nina/data_Nanopore_Isabel"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each barcode folder</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> folder_name <span class="kw">in</span> <span class="st">"</span><span class="va">$base_folder</span><span class="st">"</span>/barcode<span class="pp">*</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get the folder name (barcode01, barcode02, etc.)</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode_name</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="st">"</span><span class="va">$folder_name</span><span class="st">"</span><span class="va">)</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Concatenate all .fastq.gz files in that folder into one in the current directory</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Processing: </span><span class="va">$barcode_name</span><span class="st"> in </span><span class="va">$folder_name</span><span class="st">"</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    <span class="co">#ll "$folder_name"/*.fastq.gz</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span> <span class="st">"</span><span class="va">$folder_name</span><span class="st">"</span>/<span class="pp">*</span>.fastq.gz <span class="op">&gt;</span> data/irathmann/<span class="st">"</span><span class="va">${barcode_name}</span><span class="st">.fastq.gz"</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Created </span><span class="va">${barcode_name}</span><span class="st">.fastq.gz"</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> data/irathmann/barcode<span class="pp">*</span> <span class="op">&gt;</span> data/barcodeX.fastq.gz</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="quality-cleaning" class="level2">
<h2 class="anchored" data-anchor-id="quality-cleaning">Quality cleaning</h2>
<section id="quality-of-raw-reads" class="level3">
<h3 class="anchored" data-anchor-id="quality-of-raw-reads">Quality of raw reads</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check original quality</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span><span class="op">=</span>1 <span class="at">--mem</span><span class="op">=</span>5G seqkit stats data/barcodeX.fastq.gz  <span class="dt">\</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-Tao</span> results/seqkit/barcodeX_stats.tsv <span class="at">--threads</span> 1</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate nanoplot</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="ex">NanoPlot</span> <span class="at">--fastq</span> data/barcodeX.fastq.gz <span class="dt">\</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">-t</span> 2 <span class="dt">\</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--tsv_stats</span> <span class="dt">\</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">--plots</span> dot <span class="dt">\</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> results/nanoplot</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Notes:</p>
<ul>
<li>Seqkit:
<ul>
<li>43,170 sequences total</li>
<li>60,246,550 bp</li>
<li>Sequence length: 31 min, 1395.6 mean, 19385 max</li>
<li>Average quality: 16.63</li>
</ul></li>
<li>Nanoplot
<ul>
<li>Few reads with a phred score of 10 –&gt; will use 12 to discard sequences</li>
<li>Peak of sequence length goes from 1400 - 1500 (will use this with a 100 bp “grace area”)</li>
</ul></li>
</ul>
</section>
<section id="quality-cleaning-1" class="level3">
<h3 class="anchored" data-anchor-id="quality-cleaning-1">Quality cleaning</h3>
<section id="v1" class="level4">
<h4 class="anchored" data-anchor-id="v1">V1</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate an output folder for the results</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> results/fastplong/indiv</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Activate the fastplong conda environment</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate fastplong_0.4.1</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Run fastplong</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Ran on individual files but using stats from combined analysis</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> data/irathmann/barcode<span class="pp">*</span>fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fastq.gz<span class="va">)</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>20G <span class="at">--cpus-per-task</span> 2 fastplong <span class="dt">\</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">-i</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">-l</span> 1300 <span class="at">--length_limit</span> 1600 <span class="dt">\</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mean_qual</span> 12 <span class="at">--cut_front</span> 12 <span class="dt">\</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">--thread</span> 2 <span class="dt">\</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">-o</span> results/fastplong/indiv/<span class="va">${barcode}</span>_filtered.fastq.gz <span class="dt">\</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">--html</span> results/fastplong/indiv/<span class="va">${barcode}</span>.html <span class="dt">\</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">--json</span> results/fastplong/indiv/<span class="va">${barcode}</span>.json</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Deactivate the conda environment </span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize data </span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Output stored in: results/fastplong/indiv/fastplong_summary_v1.csv</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/parse_fastplong.py</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Notice:</p>
<p>This detected barcodes but upon closer investigation (Blastn via the webserver), the identified adapters have 100% identity to microbial strains. I did only find very few known barcode 01 sequences in the raw barcode01 fastq files.</p>
<p>Therefore, I will check with Isabel if the barcodes were already removed and also run fastplong with a Nanopore specific adapter file:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> results/fastplong/indiv/<span class="pp">*</span>fastq.gz</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> results/fastplong/indiv/<span class="pp">*</span>html</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> results/fastplong/indiv/<span class="pp">*</span>json </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Activate the fastplong conda environment</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate fastplong_0.4.1</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run fastplong</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Ran on individual files but using stats from combined analysis</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> data/irathmann/barcode<span class="pp">*</span>fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fastq.gz<span class="va">)</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>20G <span class="at">--cpus-per-task</span> 5 fastplong <span class="dt">\</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">-i</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">-l</span> 1300 <span class="at">--length_limit</span> 1600 <span class="dt">\</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mean_qual</span> 12 <span class="at">--cut_front</span> 12 <span class="dt">\</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">-a</span> /zfs/omics/projects/bioinformatics/files/nanopore/all_sequences_unique.fasta <span class="dt">\</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">--thread</span> 5 <span class="dt">\</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">-o</span> results/fastplong/indiv/<span class="va">${barcode}</span>_filtered.fastq.gz <span class="dt">\</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">--html</span> results/fastplong/indiv/<span class="va">${barcode}</span>.html <span class="dt">\</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">--json</span> results/fastplong/indiv/<span class="va">${barcode}</span>.json</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Deactivate the conda environment </span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize data </span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Output stored in: results/fastplong/indiv/fastplong_summary_v1.csv</span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/parse_fastplong.py</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>This still returns the incorrect adapters.</p>
<p>Tried to run -A with -a:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> results/fastplong/indiv/<span class="pp">*</span>fastq.gz</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> results/fastplong/indiv/<span class="pp">*</span>html</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> results/fastplong/indiv/<span class="pp">*</span>json </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Activate the fastplong conda environment</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate fastplong_0.4.1</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Run fastplong</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Ran on individual files but using stats from combined analysis</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> data/irathmann/barcode<span class="pp">*</span>fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fastq.gz<span class="va">)</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>20G <span class="at">--cpus-per-task</span> 5 fastplong <span class="dt">\</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">-i</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">-l</span> 1300 <span class="at">--length_limit</span> 1600 <span class="dt">\</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mean_qual</span> 12 <span class="at">--cut_front</span> 12 <span class="dt">\</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">-A</span> <span class="at">-a</span> /zfs/omics/projects/bioinformatics/files/nanopore/all_sequences_unique.fasta <span class="dt">\</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">--thread</span> 5 <span class="dt">\</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">-o</span> results/fastplong/indiv/<span class="va">${barcode}</span>_filtered.fastq.gz <span class="dt">\</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">--html</span> results/fastplong/indiv/<span class="va">${barcode}</span>.html <span class="dt">\</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">--json</span> results/fastplong/indiv/<span class="va">${barcode}</span>.json</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Deactivate the conda environment </span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize data </span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Output stored in: results/fastplong/indiv/fastplong_summary_v3.csv</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/parse_fastplong.py</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>This simply removes low quality and short reads but does no adapter trimming. I will check how the data was treated and for now continue working with the data that was just length and quality filtered.</p>
<table class="caption-top table">
<colgroup>
<col style="width: 15%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>barcode01</th>
<th>barcode02</th>
<th>barcode03</th>
<th>barcode04</th>
<th>barcode05</th>
<th>barcode06</th>
<th>barcode07</th>
<th>barcode09</th>
<th>barcode10</th>
<th>barcode11</th>
<th>barcode12</th>
<th>barcode13</th>
<th>barcode14</th>
<th>barcode15</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>before_total_reads</td>
<td>3919</td>
<td>2555</td>
<td>3918</td>
<td>2567</td>
<td>2261</td>
<td>2697</td>
<td>2689</td>
<td>2573</td>
<td>2240</td>
<td>2771</td>
<td>3722</td>
<td>4031</td>
<td>2975</td>
<td>4252</td>
</tr>
<tr class="even">
<td>before_total_bases</td>
<td>5466871</td>
<td>3577602</td>
<td>5507106</td>
<td>3584300</td>
<td>3192642</td>
<td>3753988</td>
<td>3779819</td>
<td>3467322</td>
<td>3099154</td>
<td>3929294</td>
<td>5277341</td>
<td>5553929</td>
<td>4131833</td>
<td>5925349</td>
</tr>
<tr class="odd">
<td>before_read_mean_length</td>
<td>1394</td>
<td>1400</td>
<td>1405</td>
<td>1396</td>
<td>1412</td>
<td>1391</td>
<td>1405</td>
<td>1347</td>
<td>1383</td>
<td>1418</td>
<td>1417</td>
<td>1377</td>
<td>1388</td>
<td>1393</td>
</tr>
<tr class="even">
<td>before_gc_content</td>
<td>0.548902</td>
<td>0.521424</td>
<td>0.532478</td>
<td>0.549224</td>
<td>0.499368</td>
<td>0.546522</td>
<td>0.546717</td>
<td>0.553473</td>
<td>0.54838</td>
<td>0.529229</td>
<td>0.529152</td>
<td>0.534887</td>
<td>0.53684</td>
<td>0.536866</td>
</tr>
<tr class="odd">
<td>after_total_reads</td>
<td>3323</td>
<td>2201</td>
<td>3402</td>
<td>2184</td>
<td>2009</td>
<td>2340</td>
<td>2335</td>
<td>2021</td>
<td>1996</td>
<td>2468</td>
<td>3293</td>
<td>3336</td>
<td>2571</td>
<td>3684</td>
</tr>
<tr class="even">
<td>after_total_bases</td>
<td>4839376</td>
<td>3175418</td>
<td>4915549</td>
<td>3179512</td>
<td>2872882</td>
<td>3388062</td>
<td>3380558</td>
<td>2950583</td>
<td>2798809</td>
<td>3585071</td>
<td>4778778</td>
<td>4840158</td>
<td>3716183</td>
<td>5327234</td>
</tr>
<tr class="odd">
<td>after_read_mean_length</td>
<td>1456</td>
<td>1442</td>
<td>1444</td>
<td>1455</td>
<td>1430</td>
<td>1447</td>
<td>1447</td>
<td>1459</td>
<td>1402</td>
<td>1452</td>
<td>1451</td>
<td>1450</td>
<td>1445</td>
<td>1446</td>
</tr>
<tr class="even">
<td>after_gc_content</td>
<td>0.549346</td>
<td>0.521657</td>
<td>0.532693</td>
<td>0.549453</td>
<td>0.499533</td>
<td>0.546727</td>
<td>0.547006</td>
<td>0.553425</td>
<td>0.548483</td>
<td>0.529471</td>
<td>0.52941</td>
<td>0.535228</td>
<td>0.537033</td>
<td>0.537187</td>
</tr>
<tr class="odd">
<td>passed_filter_reads</td>
<td>3323</td>
<td>2201</td>
<td>3402</td>
<td>2184</td>
<td>2009</td>
<td>2340</td>
<td>2335</td>
<td>2021</td>
<td>1996</td>
<td>2468</td>
<td>3293</td>
<td>3336</td>
<td>2571</td>
<td>3684</td>
</tr>
<tr class="even">
<td>low_quality_reads</td>
<td>122</td>
<td>134</td>
<td>156</td>
<td>82</td>
<td>130</td>
<td>83</td>
<td>97</td>
<td>101</td>
<td>99</td>
<td>106</td>
<td>161</td>
<td>146</td>
<td>75</td>
<td>130</td>
</tr>
<tr class="odd">
<td>too_many_N_reads</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>too_short_reads</td>
<td>454</td>
<td>204</td>
<td>338</td>
<td>287</td>
<td>114</td>
<td>263</td>
<td>246</td>
<td>438</td>
<td>139</td>
<td>192</td>
<td>256</td>
<td>531</td>
<td>315</td>
<td>422</td>
</tr>
<tr class="odd">
<td>too_long_reads</td>
<td>20</td>
<td>16</td>
<td>22</td>
<td>14</td>
<td>8</td>
<td>11</td>
<td>11</td>
<td>13</td>
<td>6</td>
<td>5</td>
<td>12</td>
<td>18</td>
<td>14</td>
<td>16</td>
</tr>
</tbody>
</table>
<p>Additionally, I want to create a more stringent dataset to see how this affects the filtering. Note this was done after testing NGSpecies ID from v0-v14.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run fastplong</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Ran on individual files but using stats from combined analysis</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate fastplong_0.4.1</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> results/fastplong/indiv/p15/</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> data/irathmann/barcode<span class="pp">*</span>fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fastq.gz<span class="va">)</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--mem</span><span class="op">=</span>20G <span class="at">--cpus-per-task</span> 5 fastplong <span class="dt">\</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">-i</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">-l</span> 1300 <span class="at">--length_limit</span> 1600 <span class="dt">\</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mean_qual</span> 15 <span class="dt">\</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">-A</span> <span class="at">-a</span> /zfs/omics/projects/bioinformatics/files/nanopore/all_sequences_unique.fasta <span class="dt">\</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">--thread</span> 5 <span class="dt">\</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">-o</span> results/fastplong/indiv/p15/<span class="va">${barcode}</span>_filtered.fastq.gz <span class="dt">\</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">--html</span> results/fastplong/indiv/p15/<span class="va">${barcode}</span>.html <span class="dt">\</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">--json</span> results/fastplong/indiv/p15/<span class="va">${barcode}</span>.json</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Deactivate the conda environment </span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> stats results/fastplong/indiv/p15/barcode15_filtered.fastq.gz <span class="dt">\</span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">-Tao</span> results/seqkit/barcode15_test.tsv</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate nanoplot</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a><span class="ex">NanoPlot</span> <span class="at">--fastq</span> results/fastplong/indiv/p15/barcode15_filtered.fastq.gz <span class="dt">\</span></span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">-t</span> 2 <span class="dt">\</span></span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">--tsv_stats</span> <span class="dt">\</span></span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">--plots</span> dot <span class="dt">\</span></span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> results/nanoplot/testing</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize data </span></span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Output stored in: results/fastplong/indiv/fastplong_summary_v4.csv</span></span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/parse_fastplong.py</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>This somehow removes less reads than expected, so lets test chopper:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate chopper0.8.0</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="ex">results/chopper</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> data/irathmann/barcode<span class="pp">*</span>fastq.gz<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$file</span> .fastq.gz<span class="va">)</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">gunzip</span> <span class="at">-c</span> <span class="va">$file</span> <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>        <span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G chopper <span class="dt">\</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">-q</span> 16 <span class="dt">\</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">-l</span> 1300 <span class="at">--maxlength</span> 1600 <span class="dt">\</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">--threads</span> 20 <span class="kw">|</span><span class="dt">\</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>        <span class="fu">gzip</span> <span class="op">&gt;</span> results/chopper/<span class="va">${barcode}</span>.fastq.gz</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Kept 2894 reads out of 4252 reads, white this was 3674 with fastplong.</p>
<p>I did some reading on how average quality scores are calculated and I think the discrepancy comes from fastplong likely using the arithmetic mean and tools like Nanoplot converting PHRED scores to error probabilities, averaging those, and converting back to a PHRED score (<a href="https://github.com/wdecoster/NanoPlot/blob/master/scripts/fastq_to_tsv.py">code</a>, <a href="https://gigabaseorgigabyte.wordpress.com/2017/06/26/averaging-basecall-quality-scores-the-right-way/">discussion</a>).</p>
<p>I tried to mimick this in two python scripts and this indeed shows differences:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mean quality over all reads: 18.57</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/my_math_nanoplot.py data/irathmann/barcode15.fastq.gz</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Mean quality over all reads: 34.71</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/my_math_fastplong.py data/irathmann/barcode15.fastq.gz</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># seqkit gives 16.83</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> stats data/irathmann/barcode15.fastq.gz <span class="at">-Tao</span> results/test.txt</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="co"># gives mean 16.8 and median 19.0</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate nanoplot</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="ex">NanoPlot</span> <span class="at">--fastq</span> data/irathmann/barcode15.fastq.gz <span class="dt">\</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">-t</span> 5 <span class="dt">\</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">--tsv_stats</span> <span class="dt">\</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">--plots</span> dot <span class="dt">\</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> results/nanoplot/testing</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The first gives a mean quality over all reads of 18.57, the second of 34.71 (checking the HTMLs of the fastplong output also seems to confirm the unusually high mean quality scores).</p>
<p>I added my observations in a currently open <a href="https://github.com/OpenGene/fastplong/issues/27">fastplong issue</a>.</p>
<p>Minimal R example:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Function to convert FASTQ quality line (ASCII) to numeric Phred scores</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>fastq_to_phred <span class="ot">&lt;-</span> <span class="cf">function</span>(qual_line, <span class="at">offset =</span> <span class="dv">33</span>) {</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># as.integer() converts the raw hex value to decimal</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># offset = 33 for Sanger/Illumina 1.8+ format</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.integer</span>(<span class="fu">charToRaw</span>(qual_line)) <span class="sc">-</span> offset</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate Arithmetic Mean</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>arith_mean <span class="ot">&lt;-</span> <span class="cf">function</span>(q) <span class="fu">mean</span>(q)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate Phred-aware Mean</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="co"># arithmetic averaging of Q scores is inappropriate because Q is logarithmic</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a><span class="co"># here: transform Q to linear probabilities before calculating the mean</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>phred_mean <span class="ot">&lt;-</span> <span class="cf">function</span>(q) {</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>    error_prop <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">^</span>(<span class="sc">-</span>q <span class="sc">/</span> <span class="dv">10</span>)</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>    avg_error <span class="ot">&lt;-</span> <span class="fu">mean</span>(error_prop)</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>    avg_phred <span class="ot">&lt;-</span> <span class="sc">-</span><span class="dv">10</span> <span class="sc">*</span> <span class="fu">log10</span>(avg_error)</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(avg_phred)</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate quality lines for three random reads representing Illumina/Nanopore data</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>qual_lines <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"IGD"</span>, <span class="st">"BCA"</span>, <span class="st">"?0&amp;"</span>)</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>read_names <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"read"</span>, <span class="fu">seq_along</span>(qual_lines))</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>reads <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">lapply</span>(qual_lines, fastq_to_phred), read_names)</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Df with all results </span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(<span class="fu">names</span>(reads), <span class="cf">function</span>(i) {</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a>    q <span class="ot">&lt;-</span> reads[[i]]</span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb9-33"><a href="#cb9-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">read =</span> i,</span>
<span id="cb9-34"><a href="#cb9-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">raw_read =</span> <span class="fu">paste</span>(q, <span class="at">collapse =</span> <span class="st">","</span>),</span>
<span id="cb9-35"><a href="#cb9-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">arithmetic =</span> <span class="fu">arith_mean</span>(q),</span>
<span id="cb9-36"><a href="#cb9-36" aria-hidden="true" tabindex="-1"></a>        <span class="at">phred_aware =</span> <span class="fu">phred_mean</span>(q)</span>
<span id="cb9-37"><a href="#cb9-37" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb9-38"><a href="#cb9-38" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb9-39"><a href="#cb9-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-40"><a href="#cb9-40" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(df, <span class="at">row.names =</span> F)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  read raw_read arithmetic phred_aware
 read1 40,38,35   37.66667   37.176671
 read2 33,34,32   33.00000   32.923584
 read3  30,15,5   16.66667    9.344819</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="clustering-with-ngspeciesid" class="level2">
<h2 class="anchored" data-anchor-id="clustering-with-ngspeciesid">Clustering with NGSpeciesID</h2>
<p>Install newest version:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Installation</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-p</span> /zfs/omics/projects/bioinformatics/software/miniconda3/envs/NGSpeciesID_0.3.1  python=3.11 pip</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate NGSpeciesID_0.3.1</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> install <span class="at">--yes</span> <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda medaka==2.0.1 openblas==0.3.3 spoa racon minimap2  samtools</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install NGSpeciesID</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Test installation </span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> test_ngspeciesID</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> test_ngspeciesID</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-LO</span> https://raw.githubusercontent.com/ksahlin/NGSpeciesID/master/test/sample_h1.fastq</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="ex">NGSpeciesID</span> <span class="at">--ont</span> <span class="at">--fastq</span> sample_h1.fastq <span class="at">--outfolder</span> ./sample_h1 <span class="at">--consensus</span> <span class="at">--medaka</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ..</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> test_ngspeciesID</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Installation vsearch </span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">-p</span> /zfs/omics/projects/bioinformatics/software/miniconda3/envs/vsearch_2.30.1  <span class="at">-c</span> bioconda vsearch==2.30.1</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="run-v0-defaults" class="level3">
<h3 class="anchored" data-anchor-id="run-v0-defaults">Run v0 defaults</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine data </span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> results/fastplong/indiv/<span class="pp">*</span>fastq.gz <span class="op">&gt;</span> results/fastplong/barcodeX.fastq.gz</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># And uncompress (NGSpeciesID takes not compressed input)</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">gzip</span> <span class="at">-d</span> results/fastplong/barcodeX.fastq.gz</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate NGSpeciesID_0.3.1 </span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># defaults aligned thres of 0.4, abundance ratio of 0.1, mapped thres of 0.7, rc identity of 0.9</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v0  <span class="dt">\</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v0b  <span class="dt">\</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01</span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v0b/consensus_reference_<span class="pp">*</span>.fasta </span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>v0:</p>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 11 clusters formed</li>
<li>Finished Consensus creation: 1 created</li>
</ul>
<p>v0b:</p>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 11 clusters formed</li>
<li>Finished Consensus creation: 5 created (one super cluster with 29090 reads)</li>
</ul>
<p>Conclusion: Balance abundance_ratio based on the number of reads generated per barcod. I.e. Considering a min. barcode sequencing depth of 2,500 reads and combining all barcodes gives ~37,000 reads –&gt; using a 0.01 cutoff should ensure that such a cluster still is found</p>
</section>
<section id="run-v1" class="level3">
<h3 class="anchored" data-anchor-id="run-v1">Run v1</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate NGSpeciesID_0.3.1 </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v1 <span class="dt">\</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.8 <span class="dt">\</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Check supporting reads (from 1937 to 9312)</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Default thresholds used by the tool:</p>
<ul>
<li><code>--aligned_threshold</code> uses 0.4</li>
<li><code>--abundance_ratio</code> uses 0.01</li>
<li><code>--mapped_threshold</code> uses 0.7</li>
</ul>
<p>Notes from run:</p>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 166 clusters formed</li>
<li>Finished Consensus creation: 8 created and being supported by 35,521 / 37,163 reads total</li>
</ul>
<p>Since some clusters are rather large, this suggest that here species are potentially fused. To test this, I will test different quality thresholds.</p>
</section>
<section id="run-v2-increase-aligned_threshold" class="level3">
<h3 class="anchored" data-anchor-id="run-v2-increase-aligned_threshold">Run v2: Increase aligned_threshold</h3>
<p>Decisions:</p>
<ul>
<li>Increase <code>-aligned_threshold</code> to 0.95. This raises the minimum alignment identity used when deciding read similarity, so near-identical but distinct 16S variants are less likely to collapse</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v2 <span class="dt">\</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Check consensus reads </span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Still 3 clusters with potentially fusions:</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. &gt;consensus_cl_id_32_total_supporting_reads_8025</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v2/consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v2/consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Notes from run:</p>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 696 clusters formed</li>
<li>Finished Consensus creation: 9 created and being supported by 30,057 / 37,163 reads total</li>
</ul>
<p>WeblastN against 16S db:</p>
<ul>
<li>consensus_cl_id_103_total_supporting_reads_3790: Huaxiibacter chinensis</li>
<li>consensus_cl_id_135_total_supporting_reads_4276: Acinetobacter guilloui</li>
<li>consensus_cl_id_139_total_supporting_reads_1706: Stenotrophomonas lactitubi</li>
<li>consensus_cl_id_182_total_supporting_reads_1722: Sphingobacterium siyangense</li>
<li>consensus_cl_id_200_total_supporting_reads_2504 : Comamonas piscis</li>
<li>consensus_cl_id_2723_total_supporting_reads_1639: Sphingomonas desiccabilis</li>
<li>consensus_cl_id_2_total_supporting_reads_4886: Enterobacter quasihormaechei</li>
<li>consensus_cl_id_32_total_supporting_reads_8025: Ectopseudomonas oleovorans</li>
<li>consensus_cl_id_713_total_supporting_reads_1509: Chryseobacterium mulctrae</li>
</ul>
<p>Next, I will increase <code>rc_identity_threshold</code> (default 0.9) the same or a bit lower (to account for error profiles/uneven trimming) to the value of <code>aligned_threshold</code>. <code>aligned_threshold</code> should control similarity in reads aligning in the same orientation and <code>rc_identity_threshold</code> should control how tolerant NGSpeciesID is when aligning a read in the reverse-complement orientation to an existing cluster representative.</p>
</section>
<section id="v3-increase-rc_identity_threshold" class="level3">
<h3 class="anchored" data-anchor-id="v3-increase-rc_identity_threshold">V3: Increase rc_identity_threshold</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v3 <span class="dt">\</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--rc_identity_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check consensus reads </span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Still 3 clusters with potentially fusions:</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. &gt;consensus_cl_id_32_total_supporting_reads_8025</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v3/consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v3/consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Notes from run:</p>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 696 clusters formed</li>
<li>Finished Consensus creation: 14 created and being supported by 30,057 / 37,163 reads total</li>
</ul>
<p>We still have one sequences that might consist of 2 taxa (consensus_cl_id_32_total_supporting_reads_4996) since the max</p>
<p>Next, I will slowly increase <code>aligned_threshold</code> but I need to keep in mind that this might split genuine species clusters if intragenomic 16S variants exist –&gt; If we see many new tiny clusters that BLAST to the same species, you have over-split.</p>
</section>
<section id="run-v4-individual" class="level3">
<h3 class="anchored" data-anchor-id="run-v4-individual">Run v4: individual</h3>
<p>Next, I will run the analysis on the individual barcodes to get a better idea what is in each individual file (since we know each barcode = one species):</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate NGSpeciesID_0.3.1 </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Unzip individual files</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gzip</span> <span class="at">-d</span> results/fastplong/indiv/<span class="pp">*</span>fastq.gz</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> results/fastplong/indiv/<span class="pp">*</span>fastq<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$file</span> .fastq <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_filtered//g'</span><span class="kw">`</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Now running </span><span class="va">${barcode}</span><span class="st">"</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">--fastq</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">--outfolder</span> results/ngspeciesid/v4/<span class="va">${barcode}</span> <span class="dt">\</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">--aligned_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">--rc_identity_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Check identity of split sequences </span></span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Low % identify but likely due to indel/deletions which are very typical in nanopore data</span></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate vsearch_2.30.1</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> results/ngspeciesid/v4/barcode04/medaka_<span class="pp">*</span>/consensus.fasta <span class="op">&gt;</span> results/ngspeciesid/v4/bc4_consensus.fasta</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Check identity of split sequences </span></span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a><span class="co"># BC04: Identity:     885/2069 (42.8%), Gaps:         937/2069 (45.3%</span></span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a><span class="ex">needle</span> <span class="at">-asequence</span> results/ngspeciesid/v4/barcode04/medaka_cl_id_0/consensus.fasta <span class="dt">\</span></span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">-bsequence</span> results/ngspeciesid/v4/barcode04/medaka_cl_id_1/consensus.fasta <span class="dt">\</span></span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">-gapopen</span> 10 <span class="at">-gapextend</span> 0.5 <span class="dt">\</span></span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">-outfile</span> results/ngspeciesid/v4/bc04_needle.txt</span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a><span class="ex">vsearch</span> <span class="at">--allpairs_global</span> results/ngspeciesid/v4/bc4_consensus.fasta <span class="dt">\</span></span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a>        <span class="at">--acceptall</span> <span class="dt">\</span></span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a>        <span class="at">--uc</span> results/ngspeciesid/v4//bc04_allpairs.uc <span class="dt">\</span></span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>        <span class="at">--iddef</span> 1 <span class="dt">\</span></span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>        <span class="at">--threads</span> 4</span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Notes:</p>
<ul>
<li>barcode 01: 1 consensus, 2813 supporting reads, Enterobacter quasihormaechei</li>
<li>**barcode 02: 2 consensus, 789 + 894 supporting reads, 2x Sphingobacterium siyangens</li>
<li>barcode 03: 1 consensus, 2393 supporting reads, Comamonas piscis</li>
<li>**barcode 04: 2 consensus, 856 +1051 supporting reads, 2x Enterobacter quasihormaechei &lt;– risk of fusing with barcode 01</li>
<li>barcode 05: 1 consensus, 1530 supporting reads, Chryseobacterium mulctrae</li>
<li>barcode 06: 1 consensus, 1931 supporting reads, Huaxiibacter chinensis</li>
<li>barcode 07: 1 consensus, 1886 supporting reads, Huaxiibacter chinensis &lt;– risk of fusion with barcode 06</li>
<li>barcode 08: does not exist</li>
<li>barcode 09: 1 consensus, 1701 supporting reads, Stenotrophomonas lactitubi</li>
<li>**barcode 10: 2 consensus, 845 + 825 supporting reads, 2x Sphingomonas desiccabilis</li>
<li>barcode 11: 1 consensus, 2077 supporting reads, Acinetobacter guillouiae</li>
<li>barcode 12: 1 consensus, 2533 supporting reads, Acinetobacter guillouiae &lt;– risk of fusion with bc11</li>
<li>barcode 13: 1 consensus, 3048 supporting reads, Pseudomonas lurida (not found in v2 blast but there was a potentially fused pseudomonas that also is part of bc14 and 15)</li>
<li>barcode 14: 1 consensus, 2139 supporting reads, Ectopseudomonas oleovorans</li>
<li>barcode 15: 1 consensus, 3083 supporting reads, Ectopseudomonas oleovorans</li>
</ul>
<p>Current summary:</p>
<ul>
<li>v2 (aligned tres 0.95, rc_identify at 0.9 (default)): 9 taxa, 3-4 consensues likely merged (clusters that likely contain multiple species)</li>
<li>v3 (aligned tres 0.95, rc_identify at 0.95): 14 taxa, 1 potentially fused consensus, 4 consensuses likely oversplit, i.e., one species split into multiple consensuses</li>
<li>Reduce rc setting to see what this does</li>
</ul>
</section>
<section id="run-v5-individuallower-rc" class="level3">
<h3 class="anchored" data-anchor-id="run-v5-individuallower-rc">Run v5: individual+lower rc</h3>
<p>Next, I will run the analysis on the individual barcodes to get a better idea what is in each individual file (since we know each barcode = one species):</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate NGSpeciesID_0.3.1 </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> results/fastplong/indiv/<span class="pp">*</span>fastq<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$file</span> .fastq <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_filtered//g'</span><span class="kw">`</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Now running </span><span class="va">${barcode}</span><span class="st">"</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">--fastq</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">--outfolder</span> results/ngspeciesid/v5/<span class="va">${barcode}</span> <span class="dt">\</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">--aligned_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">--rc_identity_threshold</span> 0.94 <span class="dt">\</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="co"># count forward vs reverse for consensus 0: 1043 fwd, 953 rc</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a><span class="ex">minimap2</span> <span class="at">-ax</span> map-ont results/ngspeciesid/v5/barcode10/consensus_reference_1.fasta <span class="dt">\</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>    results/fastplong/indiv/barcode10_filtered.fastq <span class="dt">\</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>    <span class="kw">|</span> <span class="ex">samtools</span> view <span class="at">-F4</span> <span class="at">-</span> <span class="dt">\</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>    <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{if (and($2,16)) print "RC"; else print "FWD"}'</span> <span class="dt">\</span></span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>    <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="at">-c</span></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a><span class="co"># and for consensus 1: 953 FWD, 1043 RC</span></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a><span class="ex">minimap2</span> <span class="at">-ax</span> map-ont results/ngspeciesid/v5/barcode10/consensus_reference_0.fasta <span class="dt">\</span></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>    results/fastplong/indiv/barcode10_filtered.fastq <span class="dt">\</span></span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>    <span class="kw">|</span> <span class="ex">samtools</span> view <span class="at">-F4</span> <span class="at">-</span> <span class="dt">\</span></span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>    <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{if (and($2,16)) print "RC"; else print "FWD"}'</span> <span class="dt">\</span></span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a>    <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> <span class="at">-c</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Split barcodes:</p>
<ul>
<li>barcode02 –&gt; 1 cluster, 1683 supporting reads (ok now)</li>
<li>barcode04: 2 clusters, 49 + 1858 supporting reads (better)</li>
<li>barcode 10: 2 clusters, 845 + 825 supporting reads (same)</li>
</ul>
<p>Minimap numbers suggest a reverse-complement splitting artifact. Each consensus (0 and 1) captures roughly half of the reads, and the read orientation is flipped between them (one is ~1000 FWD / 950 RC, the other the reverse). –&gt; try lowering the rc threshold</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> results/fastplong/indiv/<span class="pp">*</span>fastq<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$file</span> .fastq <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_filtered//g'</span><span class="kw">`</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Now running </span><span class="va">${barcode}</span><span class="st">"</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">--fastq</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">--outfolder</span> results/ngspeciesid/v6/<span class="va">${barcode}</span> <span class="dt">\</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">--aligned_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">--rc_identity_threshold</span> 0.93 <span class="dt">\</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/summarize_ngspecies.py <span class="at">-f</span> results/ngspeciesid/</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Split barcodes:</p>
<ul>
<li>barcode02: still ok</li>
<li>barcode04: 2 clusters, 49 + 1858 supporting reads (better)</li>
<li>barcode10: 2 clusters, 845 + 825 supporting reads (same)</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> results/fastplong/indiv/<span class="pp">*</span>fastq<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$file</span> .fastq <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_filtered//g'</span><span class="kw">`</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Now running </span><span class="va">${barcode}</span><span class="st">"</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">--fastq</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">--outfolder</span> results/ngspeciesid/v7/<span class="va">${barcode}</span> <span class="dt">\</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">--aligned_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">--rc_identity_threshold</span> 0.92 <span class="dt">\</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/summarize_ngspecies.py <span class="at">-f</span> results/ngspeciesid/</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Split barcodes:</p>
<ul>
<li>barcode02: still ok</li>
<li>barcode04: 2 clusters, 49 + 1858 supporting reads (better)</li>
<li>barcode10: 2 clusters, 845 + 825 supporting reads (same)</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> results/fastplong/indiv/<span class="pp">*</span>fastq<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>    <span class="va">barcode</span><span class="op">=</span><span class="kw">`</span><span class="fu">basename</span> <span class="va">$file</span> .fastq <span class="kw">|</span> <span class="fu">sed</span> <span class="st">'s/_filtered//g'</span><span class="kw">`</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Now running </span><span class="va">${barcode}</span><span class="st">"</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">--fastq</span> <span class="va">$file</span> <span class="dt">\</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">--outfolder</span> results/ngspeciesid/v8/<span class="va">${barcode}</span> <span class="dt">\</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">--aligned_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">--rc_identity_threshold</span> 0.9 <span class="dt">\</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> scripts/summarize_ngspecies.py <span class="at">-f</span> results/ngspeciesid/</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Now each barcodes resolves into one consensus seq:</p>
<ul>
<li>v8 barcode01 5 2813 results/ngspeciesid/v8/barcode01/consensus_reference_5.fasta</li>
<li>v8 barcode02 1 1683 results/ngspeciesid/v8/barcode02/consensus_reference_1.fasta</li>
<li>v8 barcode03 0 2393 results/ngspeciesid/v8/barcode03/consensus_reference_0.fasta</li>
<li>v8 barcode04 1 1907 results/ngspeciesid/v8/barcode04/consensus_reference_1.fasta</li>
<li>v8 barcode05 0 1530 results/ngspeciesid/v8/barcode05/consensus_reference_0.fasta</li>
<li>v8 barcode06 15 1931 results/ngspeciesid/v8/barcode06/consensus_reference_15.fasta</li>
<li>v8 barcode07 2 1886 results/ngspeciesid/v8/barcode07/consensus_reference_2.fasta</li>
<li>v8 barcode09 3 1701 results/ngspeciesid/v8/barcode09/consensus_reference_3.fasta</li>
<li>v8 barcode10 0 1670 results/ngspeciesid/v8/barcode10/consensus_reference_0.fasta</li>
<li>v8 barcode11 0 2077 results/ngspeciesid/v8/barcode11/consensus_reference_0.fasta</li>
<li>v8 barcode12 5 2553 results/ngspeciesid/v8/barcode12/consensus_reference_5.fasta</li>
<li>v8 barcode13 6 3048 results/ngspeciesid/v8/barcode13/consensus_reference_6.fasta</li>
<li>v8 barcode14 0 2139 results/ngspeciesid/v8/barcode14/consensus_reference_0.fasta</li>
<li>v8 barcode15 0 3083 results/ngspeciesid/v8/barcode15/consensus_reference_0.fasta</li>
</ul>
<p>*<strong>Important</strong>: lowering <code>--rc_identity_threshold</code> from 0.94 to 0.90 collapses those orientation-split consensuses for bc2, bc4 and bc10</p>
</section>
<section id="run-v9-adjust-aln-threshold" class="level3">
<h3 class="anchored" data-anchor-id="run-v9-adjust-aln-threshold">Run v9: adjust aln threshold</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v9 <span class="dt">\</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.96 <span class="dt">\</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--rc_identity_threshold</span> 0.90 <span class="dt">\</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check consensus reads </span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Still 3 clusters with potentially fusions:</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. &gt;consensus_cl_id_32_total_supporting_reads_8025</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v9/consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v9/consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 696 clusters formed</li>
<li>Finished Consensus creation: 9 created and being supported by 30,057 / 37,163 reads total</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v10 <span class="dt">\</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.97 <span class="dt">\</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--rc_identity_threshold</span> 0.90 <span class="dt">\</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check consensus reads </span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Still 3 clusters with potentially fusions:</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. &gt;consensus_cl_id_32_total_supporting_reads_8025</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v10/consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v10/consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 696 clusters formed</li>
<li>Finished Consensus creation: 9 created and being supported by 30,057 / 37,163 reads total</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v11 <span class="dt">\</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.98 <span class="dt">\</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--rc_identity_threshold</span> 0.90 <span class="dt">\</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check consensus reads </span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Still 3 clusters with potentially fusions:</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. &gt;consensus_cl_id_32_total_supporting_reads_8025</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v11/consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v11/consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 696 clusters formed</li>
<li>Finished Consensus creation: 9 created and being supported by 30,057 / 37,163 reads total</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v12 <span class="dt">\</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.99 <span class="dt">\</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--rc_identity_threshold</span> 0.90 <span class="dt">\</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check consensus reads </span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Still 3 clusters with potentially fusions:</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. &gt;consensus_cl_id_32_total_supporting_reads_8025</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v12/consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v12/consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 696 clusters formed</li>
<li>Finished Consensus creation: 9 created and being supported by 30,057 / 37,163 reads total</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v13 <span class="dt">\</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.997 <span class="dt">\</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--rc_identity_threshold</span> 0.90 <span class="dt">\</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.7</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check consensus reads </span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Still 3 clusters with potentially fusions:</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. &gt;consensus_cl_id_32_total_supporting_reads_8025</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v13/consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v13/consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 696 clusters formed</li>
<li>Finished Consensus creation: 9 created and being supported by 30,057 / 37,163 reads total</li>
</ul>
<p><strong>Important</strong>: <code>--aligned_threshold</code> 0.95 - 0.997 give the same results. Going from 0.94 to 0.95 changes from 8 to 9 clusters.</p>
</section>
<section id="v14-test-map-threshold" class="level3">
<h3 class="anchored" data-anchor-id="v14-test-map-threshold">v14: test map threshold</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb26"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="ex">srun</span> <span class="at">--cpus-per-task</span> 20 <span class="at">--mem</span><span class="op">=</span>50G NGSpeciesID <span class="dt">\</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--ont</span> <span class="dt">\</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fastq</span> results/fastplong/barcodeX.fastq <span class="dt">\</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outfolder</span> results/ngspeciesid/v14 <span class="dt">\</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--consensus</span> <span class="at">--medaka</span> <span class="at">--t</span> 20 <span class="dt">\</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--aligned_threshold</span> 0.95 <span class="dt">\</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--rc_identity_threshold</span> 0.90 <span class="dt">\</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--abundance_ratio</span> 0.01 <span class="dt">\</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--mapped_threshold</span> 0.55</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check consensus reads </span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Still 3 clusters with potentially fusions:</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a><span class="co"># i.e. &gt;consensus_cl_id_32_total_supporting_reads_8025</span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v14/consensus_reference_<span class="pp">*</span>.fasta </span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a><span class="co"># The reads are supported by 35,521 / 37,163 reads total </span></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"&gt;"</span> results/ngspeciesid/v14/consensus_reference_<span class="pp">*</span>.fasta  <span class="kw">|</span> <span class="fu">cut</span> <span class="at">-f10</span> <span class="at">-d</span> <span class="st">"_"</span> <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{sum += $1} END {print sum}'</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<ul>
<li>Starting Clustering: 37163 reads</li>
<li>Finished Clustering: 54 clusters formed</li>
<li>Finished Consensus creation: 8 created and being supported by 36,604 / 37,163 reads total</li>
</ul>
</section>
</section>
<section id="summary-thoughts" class="level2">
<h2 class="anchored" data-anchor-id="summary-thoughts">Summary thoughts</h2>
<p>NGSpeciesID merges clusters based on <code>--aligned_threshold</code></p>
<p>With nanopore errors (~1–2%), raising –aligned_threshold to &gt;0.995 likely won’t help, because the sequencing errors mask the small differences between closely related taxa.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>Q <span class="ot">&lt;-</span> <span class="dv">15</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>P_error <span class="ot">&lt;-</span> <span class="dv">10</span><span class="sc">^</span>(<span class="sc">-</span>Q<span class="sc">/</span><span class="dv">10</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>percent_error <span class="ot">&lt;-</span> P_error <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>percent_error</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>